### PR TITLE
[Feature](materialized view) support query match mv with agg_state on nereids planner

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes

support query match mv with agg_state on nereids planner

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.45 seconds
 stream load tsv:          453 seconds loaded 74807831229 Bytes, about 157 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          69.9 seconds inserted 10000000 Rows, about 143K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230621125133_clickbench_pr_166242.html

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.51 seconds
 stream load tsv:          458 seconds loaded 74807831229 Bytes, about 155 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          68.4 seconds inserted 10000000 Rows, about 146K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230625101715_clickbench_pr_167124.html

**Comment by BiteTheDDDDt** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java [hunk 1]
```diff
@@ -251,6 +252,21 @@ public static Optional<Slot> extractSlotOrCastOnSlot(Expression expr) {
         }
     }
 
+    /**
+     * get all slot child of expr
+     */
+    public static List<Slot> getAllSlot(Expression expr) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/ExpressionUtils.java:258]: u could use `Expression.collect()` directly: `expr.collect(SlotRefenrece.class::isInstance)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/mv/SelectMaterializedIndexWithAggregate.java [hunk 2]
```diff
@@ -1391,6 +1395,36 @@ public Expression visitNdv(Ndv ndv, RewriteContext context) {
             }
             return ndv;
         }
+
+        /**
+         * agg(col) -> agg_merge(mva_generic_aggregation__agg_state(col)) eg: max_by(k2,
+         * k3) -> max_by_merge(mva_generic_aggregation__max_by_state(k2, k3))
+         */
+        @Override
+        public Expression visitAggregateFunction(AggregateFunction aggregateFunction, RewriteContext context) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/mv/SelectMaterializedIndexWithAggregate.java:1404]: we have override some agg function in this Rewriter, such as `ndv`. So, if we have a `ndv(xx)` in sql, then it cannot processed by this function by default. if u want `visitAggregateFunction` process all aggregate function, u should call this function in `visitXXX` in this Rewriter explicitly. for example
```java
public Expression visitNdv(Ndv ndv, RewriteContext context) {
    Expression newNdv = visitAggregateFunction(ndv, context);
    if (newNdv != ndv) {
        return newNdv;
    }
    ...
}

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by zhangstar333 — APPROVED** [review]

**Pull request merged**